### PR TITLE
validate the status of app version to be ready_for_review before submitting the app

### PR DESCRIPTION
**Pull request by 404pilot**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

This PR tries to fix the [issue](https://github.com/fastlane/fastlane/issues/20065).

When a specific app version is ready to be submitted for review, `fastlane` executes a series of API requests. 

``` ruby
      submission.add_app_store_version_to_review_items(app_store_version_id: version.id)
      submission.submit_for_review # it throws an error
```

It fails at [line](https://github.com/fastlane/fastlane/blob/acb7d84e7ac068e424da231c2eca39fe2ec7297d/deliver/lib/deliver/submit_for_review.rb#L50) because the state of the version is not "READY_FOR_REVIEW". Fastlane sends the API request too fast.

### Description

Basically I added a block to wait till its state becomes 'READY_FOR_REVIEW"

### Testing Steps

```
bundle exec fastlane test
```

### CODE REVIEW

**Comment by getaaron** [comment]
Thanks so much for the PR! It seems like a nice general approach to this frustrating problem. I think it would be good to make 2 small additions before merging:

- we should provide a config to opt out (similar to `skip_waiting_for_build_processing`) 
- we should use [`BuildWatcher`](https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/build_watcher.rb) since it has more advanced support for this e.g. timeouts, handling leading zeroes in version numbers, etc.

WDYT?

**Comment by 404pilot** [comment]
@getaaron thanks for the comment. 

1. what is the scenario that people want to opt out?
2. Do you think it is appropriate to create another file `VersionWatcher` and use `skip_waiting_for_version_processing`? The build is actually done at this point and it is just that Apple needs to verify the submitted version is good

**Review by thoro23 — APPROVED** [review]

**Comment by thoro23** [comment]
Great PR! We just ran into this issue and this PR would have saved us some time and nerves.

**Comment by bakatsuyuki** [comment]
When does this PR merge? I'm waiting.

**Comment by getaaron** [comment]
@404pilot actually on thinking about it more I don't think we need to provide an opt out.

**Comment by staust** [comment]
We are facing the same issue. Would like to see this merged.

**Comment by ilkap07** [comment]
What's the status of this PR? We're having the same issue.

**Comment by Stepoura** [comment]
Any update on this? We would like to see this merged :)

File: deliver/lib/deliver/submit_for_review.rb
Comment (cstaud) [reply deliver/lib/deliver/submit_for_review.rb:51]: sounds reasonable

**Comment by ilkap07** [comment]
Any updates on this? We're still having the issue. Would be nice to see this merged

**Comment by hoangcharles0** [comment]
@404pilot  Thanks for working on this change! I was wondering when will this PR get merged. I have been experiencing this issue for the past year and this would be great to automate submitting for review

**Comment by 404pilot** [comment]
> @404pilot Thanks for working on this change! I was wondering when will this PR get merged. I have been experiencing this issue for the past year and this would be great to automate submitting for review

I am still waiting for someone to review the PR. I would suggest using `spaceship` or apple connect API to release it after the release step from fastlane fails if you want to automate the workflow now

**Comment by ilkap07** [comment]
Any updates on this PR?

**Comment by 404pilot** [comment]
can one of admins review it please :)

**Comment by stonko1994** [comment]
Any updates on this? Still an issue

**Review by getaaron — APPROVED** [review]

**Comment by getaaron** [comment]
Thanks very much for the pull request 🥳

**Pull request merged**
